### ISSUE TITLE
ERROR:  variable not found in subplan target list(only planner)

### ISSUE THREAD

**Issue by charliettxx** [greenplum-db/gpdb#16117]
## Bug Report
ERROR:  variable not found in subplan target list

### Greenplum version or build
only main

### autoconf options used ( config.status --config )
set optimizer = off;

### Expected behavior
no error, executable

### Actual behavior
ERROR:  variable not found in subplan target list

### Step to reproduce the behavior
set enable_groupagg = off;
set gp_eager_two_phase_agg = on;

create table funcdep1(a int primary key, b int, c int, d int);
create table funcdep2(a int, b int, c int, d int);

insert into funcdep1 values(1,1,1,1);
insert into funcdep1 values(2,1,1,1);
insert into funcdep1 values(3,1,1,1);
insert into funcdep2 values(1,1,1,1);

explain (costs off) select count(distinct b), sum(b), c from funcdep1 group by a;

**Comment by cstarc** [comment]
after patch #16118,  I find `explain (costs off) select count(distinct b), sum(b), c from funcdep1 group by a;`  is ok, but `explain verbose  select count(distinct b) filter(where c>1), count(distinct c)filter(where d>1), count(distinct c)filter(where d>1)-count(distinct b) filter(where d>1)  from funcdep1 group by a;` is not ok, will report ERROR: variable not found in subplan target list.  after analyze, i find in fetch_multi_dqas_info, ctx->agg_partial_costs->distinctAggrefs and ctx->agg_final_costs->distinctAggrefs may have diffent length (Because agg_partial_costs->distinctAggrefs deduplicates in make_partial_grouping_target) , this results in aggref_final->aggfilter not being set to null, and eventually results in an agg mismatch. and after add code in fetch_multi_dqas_info as follows, it is ok  to execute **`explain verbose  select count(distinct b) filter(where c>1), count(distinct c)filter(where d>1), count(distinct c)filter(where d>1)-count(distinct b) filter(where d>1)  from funcdep1 group by a;`** :
```
@@ -2337,6 +2340,11 @@
 		aggref->aggfilter = NULL;
 		aggref_final->aggfilter = NULL;
 	}
+	foreach (lcc, ctx->agg_final_costs->distinctAggrefs)
+	{
+		Aggref	        *aggref_final = (Aggref *) lfirst(lcc);
+		aggref_final->aggfilter = NULL;
+	}
 	info->dNumDistinctGroups = dNumDistinctGroups;
```

**Comment by charliettxx** [comment]
> after patch #16118, I find `explain (costs off) select count(distinct b), sum(b), c from funcdep1 group by a;` is ok, but SQL:`explain verbose select count(distinct b) filter(where c>1), count(distinct c)filter(where d>1), count(distinct c)filter(where d>1)-count(distinct b) filter(where d>1) from funcdep1 group by a;` is not ok, will report "ERROR: variable not found in subplan target list". after analyze, i find in fetch_multi_dqas_info, ctx->agg_partial_costs->distinctAggrefs and ctx->agg_final_costs->distinctAggrefs may have diffent length (Because agg_partial_costs->distinctAggrefs deduplicates in make_partial_grouping_target) , this results in aggref_final->aggfilter not being set to null, and eventually results in an agg mismatch. and after add code in fetch_multi_dqas_info as follows, it is ok to execute **`explain verbose select count(distinct b) filter(where c>1), count(distinct c)filter(where d>1), count(distinct c)filter(where d>1)-count(distinct b) filter(where d>1) from funcdep1 group by a;`** :
> 
> ```
> @@ -2337,6 +2340,11 @@
>  		aggref->aggfilter = NULL;
>  		aggref_final->aggfilter = NULL;
>  	}
> +	foreach (lcc, ctx->agg_final_costs->distinctAggrefs)
> +	{
> +		Aggref	        *aggref_final = (Aggref *) lfirst(lcc);
> +		aggref_final->aggfilter = NULL;
> +	}
>  	info->dNumDistinctGroups = dNumDistinctGroups;
> ```

Thanks for your work and explanation on it. This feels like another issue exiting in main branch whether applying https://github.com/greenplum-db/gpdb/pull/16118 or not and maybe no relevant primary key issue here.
I appreciate for your solution but just confused why ctx->agg_partial_costs->distinctAggrefs and ctx->agg_final_costs->distinctAggrefs has different length, and maybe we need to keep them same.

**Comment by cstarc** [comment]
because ctx->agg_partial_costs->distinctAggrefs is actually generated by `make_partial_grouping_target(root, grouped_rel->reltarget, extra->havingQual)` in create_partial_grouping_paths, it will finally use add_new_column_to_pathtarget to add agg(deduplicates ):
```
void
add_new_column_to_pathtarget(PathTarget *target, Expr *expr)
{
	if (!list_member(target->exprs, expr))
		add_column_to_pathtarget(target, expr, 0);
}
```
but agg_final_costs->distinctAggrefs is direct from grouped_rel->reltarget without deduplicates 

```
	partially_grouped_rel->reltarget =
		make_partial_grouping_target(root, grouped_rel->reltarget,
									 extra->havingQual);

	if (!extra->partial_costs_set)
	{
		/*
		 * Collect statistics about aggregates for estimating costs of
		 * performing aggregation in parallel.
		 */
		MemSet(agg_partial_costs, 0, sizeof(AggClauseCosts));
		MemSet(agg_final_costs, 0, sizeof(AggClauseCosts));
		if (parse->hasAggs)
		{
			List	   *partial_target_exprs;

			/* partial phase */
			partial_target_exprs = partially_grouped_rel->reltarget->exprs;
			get_agg_clause_costs(root, (Node *) partial_target_exprs,
								 AGGSPLIT_INITIAL_SERIAL,
								 agg_partial_costs);

			/* final phase */
			get_agg_clause_costs(root, (Node *) grouped_rel->reltarget->exprs,
								 AGGSPLIT_FINAL_DESERIAL,
								 agg_final_costs);
			get_agg_clause_costs(root, extra->havingQual,
								 AGGSPLIT_FINAL_DESERIAL,
								 agg_final_costs);
		}

		extra->partial_costs_set = true;
	}
```

**Comment by charliettxx** [comment]
> because ctx->agg_partial_costs->distinctAggrefs is actually generated by `make_partial_grouping_target(root, grouped_rel->reltarget, extra->havingQual)` in create_partial_grouping_paths, it will finally use add_new_column_to_pathtarget to add agg(deduplicates ):
> 

make sense, actually I've tired to access another approach eliminating reduntant filters of DQA in other places, unfortunately fialed. After apply your patch I found that query results seems consistantly, So I decide to raise
another issue https://github.com/greenplum-db/gpdb/issues/16627 recording this and need more work on it.

sql:
select count(distinct b) filter(where c>1), count(distinct c)filter(where d>1), count(distinct c)filter(where d>1)-count(distinct b) filter(where d>1) from foo group by a;
```
set enable_hashagg = on (including our patch)
 count | count | ?column? 
-------+-------+----------
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
(9 rows)

set enable_hashagg = off (back to groupagg)
 count | count | ?column? 
-------+-------+----------
     0 |     0 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
     1 |     1 |        0
(10 rows)
```
